gzip/gzip.go

```diff
@@ -67,44 +69,45 @@ func (c CompressionCodec) Code() int8 {
 	return Code
 }
 
-// Encode implements the kafka.CompressionCodec interface.
-func (c CompressionCodec) Encode(src []byte) ([]byte, error) {
-	buf := bytes.Buffer{}
-	buf.Grow(len(src)) // guess a size to avoid repeat allocations.
-	writer := writerPool.Get().(*gzip.Writer)
-	writer.Reset(&buf)
-
-	_, err := writer.Write(src)
-	if err != nil {
-		// don't return writer to pool on error.
-		return nil, err
-	}
-
-	// note that the gzip reader must be closed in order for it to write
-	// out trailing contents.  Flush is insufficient.  it is okay to re-use
-	// the writer even after it's closed by Resetting it.
-	err = writer.Close()
-	if err != nil {
-		// don't return writer to pool on error.
-		return nil, err
-	}
+// Name implements the kafka.CompressionCodec interface.
+func (c CompressionCodec) Name() string {
+	return "gzip"
+}
 
-	writerPool.Put(writer)
+// NewReader implements the kafka.CompressionCodec interface.
+func (c CompressionCodec) NewReader(r io.Reader) io.ReadCloser {
+	z := readerPool.Get().(*gzip.Reader)
+	z.Reset(r)
+	return &reader{z}
+}
 
-	return buf.Bytes(), err
+// NewWriter implements the kafka.CompressionCodec interface.
+func (c CompressionCodec) NewWriter(w io.Writer) io.WriteCloser {
+	z := writerPool.Get().(*gzip.Writer)
+	z.Reset(w)
+	return &writer{z}
 }
 
-// Decode implements the kafka.CompressionCodec interface.
-func (c CompressionCodec) Decode(src []byte) ([]byte, error) {
-	reader := readerPool.Get().(*gzip.Reader)
-	err := reader.Reset(bytes.NewReader(src))
-	if err != nil {
-		return nil, err
+type reader struct{ *gzip.Reader }
+
+func (r *reader) Close() (err error) {
+	if z := r.Reader; z != nil {
+		r.Reader = nil
+		err = z.Close()
+		z.Reset(newEmptyGzipFile())
```

Comment:
Why do we reset here to the empty file?  Is it to clear out retained buffers?  Is there a cost to calling reset with an empty file knowing that we're going to reset it with actual content later?

Reply:
This is to avoid retaining the underlying io.Reader which will probably be some combination of *bufio.Reader and *net.TCPConn.